codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -254,6 +259,46 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
         }
     }
 
+    @Override
+    protected ChannelPromise encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out, ChannelPromise promise)
+            throws Exception {
+        try {
+            encode(ctx, msg, out);
+        } catch (Exception e) {
+            if (bufferPromise != null) {
+                bufferPromise.tryFailure(e);
+            }
+            throw e;
+        }
+
+        if (bufferPromise == null) {
+            if (finished) {
+                return promise;
+            }
+            if (buffer.isReadable()) {
+                bufferPromise = promise;
+                return ctx.newPromise();
+            }
+            return promise;
+        } else {
+            if (buffer.isReadable()) {
+                if (out.isReadable()) {
+                    final ChannelPromise prev = bufferPromise;
+                    bufferPromise = promise;
+                    return prev;
+                } else {
+                    bufferPromise.addListener(new ChannelPromiseNotifier(promise));
+                    return ctx.newPromise();
```

Comment:
This doesn't look right... if the returned promise will be failed we will never fail `promise` or `bufferPromise` which I think we should

Reply:
My understanding is that we retain promises till the compression is finished, upon which we complete all of them.

Reply:
I need to look into it again in detail... if this is the case please lets add comment

Reply:
I think @normanmaurer is mostly correct here. For the case when `out` is empty I don't think we have to fail the message's promise when this new promise fails (since no part of the message was in `out` right?).
But in the other scenario where part of the message is buffered and part is in `out` we have to make sure that both the buffer promise and the returned promise that is passed along with `out` when it's written succeed (conversely, failure of either should fail the message's promise).

Hope that explanation makes sense, I also added comments inline for all possible branches now and tried to extract + document the chaining of two promises into one that hopefully help here.

Let me know what you think or if I missed a case.

Thanks! :)

Reply:
BTW: I think this needs additional tests, sorry for not adding them right away but I wanted to make sure my approach is going in the correct direction first.